src/platform/macos/input.cpp

```diff
@@ -282,16 +366,99 @@ void unicode(input_t &input, char *utf8, int size) {
+    }
+
+    IOObjectRelease(virtgamepad_service);
+  }
+
+  IOObjectRelease(virtgamepad_iterator);
+
+  if(!found) {
+    BOOST_LOG(info) << "Gamepad: Unable to open IOService."sv;
+    return 1;
+  }
+
+  // Fill up the input arguments.
+  virtgamepad_input[0] = (uint64_t)strdup(VIRTGAMEPAD_NAME);   // device name
+  virtgamepad_input[1] = strlen((char *)virtgamepad_input[0]); // name length
+  virtgamepad_input[2] = (uint64_t)gamepad_report_descriptor;  // report descriptor
+  virtgamepad_input[3] = sizeof(gamepad_report_descriptor);    // report descriptor len
+  virtgamepad_input[4] = (uint64_t)strdup(VIRTGAMEPAD_SN);     // serial number
+  virtgamepad_input[5] = strlen((char *)virtgamepad_input[4]); // serial number len
+  virtgamepad_input[6] = (uint64_t)0xdead;                     // vendor ID
+  virtgamepad_input[7] = (uint64_t)0xbeef;                     // device ID
+
+  ret = IOConnectCallScalarMethod(virtgamepad_connect, FOOHID_CREATE, virtgamepad_input, VIRTGAMEPAD_INPUT_COUNT, NULL, 0);
+  if(ret != KERN_SUCCESS) {
+    BOOST_LOG(info) << "Gamepad: Unable to create HID device. May be fine if created previously."sv;
+  }
+  return 0;
 }
 
 void free_gamepad(input_t &input, int nr) {
-  BOOST_LOG(info) << "free_gamepad: Gamepad not yet implemented for MacOS."sv;
+  IOObjectRelease(virtgamepad_connect);
 }
 
 void gamepad(input_t &input, int nr, const gamepad_state_t &gamepad_state) {
-  BOOST_LOG(info) << "gamepad: Gamepad not yet implemented for MacOS."sv;
+  // Arguments to be passed through the HID message.
+  struct gamepad_report_t gamepad;
+  uint32_t send_count = 4;
+  uint64_t send[send_count];
+  send[0] = (uint64_t)virtgamepad_input[0];       // device name
+  send[1] = strlen((char *)virtgamepad_input[0]); // name length
+  send[2] = (uint64_t)&gamepad;                   // mouse struct
+  send[3] = sizeof(struct gamepad_report_t);      // mouse struct len
+
+  auto flags = gamepad_state.buttonFlags;
+
+  // reset buttons
+  gamepad.buttons = GAMEPAD_ZEROBTNS;
+
+  // Map buttons
+  if(flags & LEFT_STICK) gamepad.buttons |= GAMEPAD_BTN_LS;
+  if(flags & RIGHT_STICK) gamepad.buttons |= GAMEPAD_BTN_RS;
+  if(flags & LEFT_BUTTON) gamepad.buttons |= GAMEPAD_BTN_L;
+  if(flags & RIGHT_BUTTON) gamepad.buttons |= GAMEPAD_BTN_R;
+  if(flags & START) gamepad.buttons |= GAMEPAD_BTN_START;
+  if(flags & BACK) gamepad.buttons |= GAMEPAD_BTN_BACK;
+  if(flags & A) gamepad.buttons |= GAMEPAD_BTN_A;
+  if(flags & B) gamepad.buttons |= GAMEPAD_BTN_B;
+  if(flags & X) gamepad.buttons |= GAMEPAD_BTN_X;
+  if(flags & Y) gamepad.buttons |= GAMEPAD_BTN_Y;
+  if(flags & DPAD_UP) gamepad.buttons |= GAMEPAD_BTN_UP;
+  if(flags & DPAD_DOWN) gamepad.buttons |= GAMEPAD_BTN_DOWN;
+  if(flags & DPAD_LEFT) gamepad.buttons |= GAMEPAD_BTN_LEFT;
+  if(flags & DPAD_RIGHT) gamepad.buttons |= GAMEPAD_BTN_RIGHT;
+
+  // Map triggers
+  if(gamepad_state.lt > 0) gamepad.buttons |= GAMEPAD_BTN_LT;
+  if(gamepad_state.rt > 0) gamepad.buttons |= GAMEPAD_BTN_RT;
```

Comment:
Is it possible to emulate these with analog axes? Digital triggers are problematic for driving games and other games where precise trigger control is required.

Reply:
I'm not sure that now I can implement analog support.

Reply:
And, I don't have device with analog trigger, lol.

Reply:
Updated. See comments in PR.